### PR TITLE
Derived streams

### PR DESCRIPTION
**Pull request by mreid-moz**
Add a `get_records` function that can be used to fetch (and filter) data for derived streams using the Pipeline metadata.

### CODE REVIEW

**Comment by mreid-moz** [comment]
Testing: I ran the "Basics" and "Histograms" parts of the Telemetry Hello World notebook (with fraction=1) both before and after this change, and the outputs all look identical.

**Comment by vitillo** [comment]
Please bump up the version in setup.py.

File: moztelemetry/spark.py [hunk 1]
```diff
@@ -145,6 +150,95 @@ def get_one_ping_per_client(pings):
...
+    parallelism = max(len(sample), sc.defaultParallelism)
+    ranges = sc.parallelize(sample, parallelism).flatMap(_read_v4_ranges).collect()
+
+    if len(ranges) == 0:
+        return sc.parallelize([])
+    else:
+        return sc.parallelize(ranges, len(ranges)).flatMap(_read_v4_range)
+
+def _list_s3_filenames(bucket, prefix, schema):
+    filenames = []
```
Comment (vitillo) [inline moztelemetry/spark.py]: nit: a list comprehension here would be nice

Comment (mreid-moz) [inline moztelemetry/spark.py]: Can I pass an iterator to "random.choice" instead of an array? If so, I don't even need this function.

**Comment by vitillo** [comment]
Looks good, you should add a new section to our [Hello World](https://github.com/mozilla/emr-bootstrap-spark/blob/master/Telemetry%20Hello%20World.ipynb) tutorial to briefly explain how to access a derived stream.

File: moztelemetry/spark.py [hunk 2]
```diff
@@ -145,6 +150,92 @@ def get_one_ping_per_client(pings):
                     reduceByKey(lambda p1, p2: p1).\
                     map(lambda p: p[1])
 
+def get_data_sources():
```
Comment (vitillo) [inline moztelemetry/spark.py]: Please prefix this method with an underscore.

File: moztelemetry/spark.py [hunk 3]
```diff
@@ -145,6 +150,92 @@ def get_one_ping_per_client(pings):
...
+    if "schema" not in _sources[source_name]:
+        try:
+            key = _bucket_meta.get_key("{}/schema.json".format(_sources[source_name].get("metadata_prefix", source_name)))
+            schema = key.get_contents_as_string()
+            _sources[source_name]["schema"] = json.loads(schema)
+        except ssl.SSLError:
+            return None
+    return _sources[source_name]["schema"]
+
+def get_records(sc, source_name, **kwargs):
```
Comment (vitillo) [inline moztelemetry/spark.py]: This public function should be located at the top of the files, with the other public functions.

**Comment by mreid-moz** [comment]
I filed https://github.com/mozilla/emr-bootstrap-spark/issues/1 to update the example notebook / documentation.

File: moztelemetry/spark.py [hunk 4]
```diff
@@ -146,6 +151,118 @@ def get_one_ping_per_client(pings):
...
+
+    All data sources support
+    :param fraction: the fraction of files read from the data source, set to 1.0
+                     by default.
+
+    Depending on the data source, different filtering criteria will be available.
+
+    Filtering criteria should be specified using the TelemetrySchema approach:
+    Range filter: submissionDate={"min": "20150901", "max": "20150908"}
+                     or submissionDate=("20150901", "20150908")
```
Comment (vitillo) [inline moztelemetry/spark.py]: Please fix the indentation here.

**Pull request merged**